spark3/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java

```diff
@@ -0,0 +1,234 @@
+
+    CatalogPlugin sessionCat = spark().sessionState().catalogManager().v2SessionCatalog();
+    Identifier sourceIdent = toCatalogAndIdentifier(args.getString(1), PARAMETERS[1].name(), sessionCat).identifier();
+
+    Map<String, String> partitionFilter = Maps.newHashMap();
+    if (!args.isNullAt(2)) {
+      args.getMap(2).foreach(DataTypes.StringType, DataTypes.StringType,
+          (k, v) -> {
+            partitionFilter.put(k.toString(), v.toString());
+            return BoxedUnit.UNIT;
+          });
+    }
+
+    long addedFilesCount = importToIceberg(tableIdent, sourceIdent, partitionFilter);
+    return new InternalRow[]{newInternalRow(addedFilesCount)};
+  }
+
+  private boolean isFileIdentifier(Identifier ident) {
+    String[] namespace = ident.namespace();
+    return namespace.length == 1 &&
+        (namespace[0].equalsIgnoreCase("orc") ||
+            namespace[0].equalsIgnoreCase("parquet") ||
+            namespace[0].equalsIgnoreCase("avro"));
+  }
+
+  private long importToIceberg(Identifier destIdent, Identifier sourceIdent, Map<String, String> partitionFilter) {
+    return modifyIcebergTable(destIdent, table -> {
+
+      validatePartitionSpec(table, partitionFilter);
+      ensureNameMappingPresent(table);
+
+      if (isFileIdentifier(sourceIdent)) {
+        Path sourcePath = new Path(sourceIdent.name());
+        String format = sourceIdent.namespace()[0];
+        importFileTable(table, sourcePath, format, partitionFilter);
+      } else {
+        importCatalogTable(table, sourceIdent, partitionFilter);
+      }
+
+      Snapshot snapshot = table.currentSnapshot();
+      return Long.parseLong(snapshot.summary().get(SnapshotSummary.ADDED_FILES_PROP));
+    });
+  }
+
+  private static void ensureNameMappingPresent(Table table) {
+    if (table.properties().get(TableProperties.DEFAULT_NAME_MAPPING) == null) {
+      // Forces Name based resolution instead of position based resolution
+      NameMapping mapping = MappingUtil.create(table.schema());
+      String mappingJson = NameMappingParser.toJson(mapping);
+      table.updateProperties()
+          .set(TableProperties.DEFAULT_NAME_MAPPING, mappingJson)
+          .commit();
+    }
+  }
+
+  private void importFileTable(Table table, Path tableLocation, String format, Map<String, String> partitionFilter) {
+    // List Partitions via Spark InMemory file search interface
+    List<SparkPartition> partitions = Spark3Util.getPartitions(spark(), tableLocation, format);
+
+    if (table.spec().isUnpartitioned()) {
+      Preconditions.checkArgument(partitions.isEmpty(), "Cannot add partitioned files to an unpartitioned table");
+      // Build a Global Partition for the source
+      SparkPartition partition = new SparkPartition(Collections.emptyMap(), tableLocation.toString(), format);
+      Preconditions.checkArgument(partitionFilter.isEmpty(), "Cannot use a partition filter when importing" +
```

Comment:
nit: Shall we group Preconditions together before constructing a global partition?